### PR TITLE
Add wechat_pay source

### PR DESCRIPTION
**Pull request by AnasNaouchi**
## Description

Implement wechat_pay source. More info [here](https://docs.opn.ooo/wechat-pay-online)

### CODE REVIEW

File: src/main/java/co/omise/models/Source.java [hunk 1]
```diff
@@ -88,10 +89,16 @@ public void setCurrency(String currency) {
     public String getEmail() {
         return this.email;
     }
+    public String getIp() {
```
Comment (danfowler) [inline src/main/java/co/omise/models/Source.java:92]:
```suggestion

        public String getIp() {
```

File: src/main/java/co/omise/models/Source.java [hunk 2]
```diff
@@ -88,10 +89,16 @@ public void setCurrency(String currency) {
...
+        return this.ip;
+    }
 
     public void setEmail(String email) {
         this.email = email;
     }
+    public void setIp(String ip) {
```
Comment (danfowler) [inline src/main/java/co/omise/models/Source.java:99]:
```suggestion

    public void setIp(String ip) {
```

File: src/main/java/co/omise/models/Source.java [hunk 3]
```diff
@@ -347,6 +360,9 @@ public CreateRequestBuilder platformType(PlatformType platformType) {
 
         @Override
         protected RequestBody payload() throws IOException {
+            if (type == SourceType.WechatPay && (ip == null || ip.isEmpty())) {
```
Comment (danfowler) [inline src/main/java/co/omise/models/Source.java:363]: Maybe we should add some commentary to justify why we are doing this validation on the client-side: the server can't know whether the caller is coming from a customer device or not.

Comment (aashishgurung) [reply src/main/java/co/omise/models/Source.java:363]: This `payload()` is shared by all the APMs. Can we move this validation logic to a different class or function for now? This sets the groundwork for handling potential unique requirements in different APMs without cluttering the main code with numerous if statements.

Comment (AnasNaouchi) [reply src/main/java/co/omise/models/Source.java]: The check has been removed.

File: src/main/java/co/omise/requests/Request.java [hunk 4]
```diff
@@ -70,6 +73,22 @@ public RequestBody getPayload() {
...
+     * @return the payload as a readable string
+     */
+    public  String getPayloadToString(){
+        try {
+            final Buffer buffer = new Buffer();
+            payload.writeTo(buffer);
+            return buffer.readUtf8();
+        }
+        catch (final IOException e) {
+            return "Unable to process payload";
```
Comment (aashishgurung) [inline src/main/java/co/omise/requests/Request.java:88]: Can you help understand why are we returning a string instead of throwing an exception?

Comment (AnasNaouchi) [reply src/main/java/co/omise/requests/Request.java]: Fixed

File: src/main/java/co/omise/models/Source.java [hunk 5]
```diff
@@ -89,10 +90,18 @@ public String getEmail() {
         return this.email;
     }
 
+        public String getIp() {
```
Comment (danfowler) [inline src/main/java/co/omise/models/Source.java:93]:
```suggestion
public String getIp() {
```

Comment (danfowler) [reply src/main/java/co/omise/models/Source.java:93]: Do we have a linter that can catch these formatting issues?

File: src/main/java/co/omise/models/Source.java [hunk 6]
```diff
@@ -281,6 +292,10 @@ public CreateRequestBuilder email(String email) {
             this.email = email;
             return this;
         }
+        public CreateRequestBuilder ip(String ip) {
```
Comment (danfowler) [inline src/main/java/co/omise/models/Source.java:295]:
```suggestion

        public CreateRequestBuilder ip(String ip) {
```

File: src/main/java/co/omise/models/Source.java (hunk 5 above)
Comment (danfowler) [reply src/main/java/co/omise/models/Source.java:93]: And, oops, I think this was my fault. I blame GitHub.

File: src/main/java/co/omise/requests/Request.java [hunk 7]
```diff
@@ -70,6 +73,17 @@ public RequestBody getPayload() {
         return payload;
     }
 
+    /**
+     * Gets payload (request params).
+     *
+     * @return the payload as a readable string
+     */
+    public String getPayloadToString() throws IOException {
```
Comment (danfowler) [inline src/main/java/co/omise/requests/Request.java:81]: @AnasNaouchi If we are adding, shouldn't we use somewhere?

Comment (AnasNaouchi) [reply src/main/java/co/omise/requests/Request.java:81]: Its just a debugging function, we did not use it as we don't need to view the body in the actual code but every time a problem occurs with the payload we need to view it using this function.

Comment (danfowler) [reply src/main/java/co/omise/requests/Request.java:81]: @AnasNaouchi wouldn't it be useful in a test to verify the JSON body is what we expect?

Comment (AnasNaouchi) [reply src/main/java/co/omise/requests/Request.java:81]: It will not be what we want as we will have null values for keys hat are excluded

Comment (danfowler) [reply src/main/java/co/omise/requests/Request.java:81]: But thinking back to your confusion earlier, if you saw a test that explicitly showed how the Java integration worked, wouldn't that have been helpful?

Comment (AnasNaouchi) [reply src/main/java/co/omise/requests/Request.java:81]: Will add unit tests for that

**Review by danfowler — APPROVED** [review]

**Review by vault087 — APPROVED** [review]

**Pull request merged**